Rewrite of Trafikverket weather - Multiple sensor types supported

endor-force opened the pull request:
## Description:
Rewritten with library pytrafikverket for retreiving data.


**Breaking change, requires user to update configuration!**
Instead of having multiple sensor configurations per station, add only one configuration per station and select the type of measurement data to subscribe to using `monitored_conditions`.
The configuration value `type` should no longer be used.

Supports more weather data from Trafikverket API.
(*Text values returned from API is string in Swedish (`wind_direction_text` and `precipitation`).*)

**Related issue (if applicable):** fixes N/A

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#6006

## Example entry for `configuration.yaml` (if applicable):
```yaml
- sensor: 
   - platform: trafikverket_weatherstation
     name: Trafikverket Precipitation WeatherStation Upplid
     api_key: eXXcbXXXacXXXXc39XX3aXXX4aXX46XX
     station: Upplid
     monitored_conditions:
         - air_temp
         - road_temp
         - humidity
         - precipitation
         - wind_direction
         - wind_direction_text
         - wind_speed

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools: 
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.  **N/A**

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

endor-force committed:
Added precipitation type from API

Enables users to see type of precipitation.
Value returned from API is a string in swedish.

endor-force commented:
Moved from master to dev, tag "merging to master" can be cleared.

endor-force committed:
Corrected tox verification errors

Correction of tox findings

endor-force committed:
Missed in tox - fixed

endor-force committed:
Hound witespace fix

endor-force committed:
Updated comment to trigger travis rebuild

Travis tox failed due to problem with tox build process. 
Correcting in a comment to trigger retry in travis..

endor-force committed:
Try to retrigger travis/tox successful rebuild

endor-force committed:
Cleaning

endor-force committed:
Cleaning more

balloob commented on homeassistant/components/sensor/trafikverket_weatherstation.py:
```diff
@@ -98,7 +104,7 @@ def update(self):
                           <EQ name="Name" value='""" + self._station + """' />
                     </FILTER>
-                    <INCLUDE>Measurement.""" + air_vs_road + """.Temp</INCLUDE>
+                    <INCLUDE>Measurement.""" + air_vs_road + """.""" + keytype + """</INCLUDE>
```
We cannot accept any PR for this platform until all protocol specific code has been extracted into a third party lib which is then used.

fabaff reviewed (dismissed):
Conflict

endor-force commented:
Sorry, rookie here, do you know any sensor that I can look at for reference on putting protocol code in separate library?

fabaff commented:
`luftnetz`, `pi_hole`, `shodan`, `swiss_public_transportation`, `thinkingcleaner`, `vasttrafik` and some others.

MartinHjelmare commented:
Maybe try to extend pytrafikverket:
https://github.com/AnderssonPeter/pytrafikverket

endor-force commented:
Sure, I will investigate the possibilities of using and extending pytrafikverket for this.

endor-force committed:
Trafikverket rebuilt for library

Extended pytrafikverket with weather sensor collction
Changed behaviour of sensor component to use pytrafikverket.
Added more sensors.

User need to change config to use new version.
 [] Documentation needs to be updated

endor-force committed:
Merge branch 'dev' into update_trafikverketsensor

MartinHjelmare commented on homeassistant/components/sensor/trafikverket_weatherstation.py:
```diff
@@ -4,119 +4,117 @@
     """Set up the Trafikverket sensor platform."""
+    from pytrafikverket.trafikverket_weather import TrafikverketWeather
+
     sensor_name = config.get(CONF_NAME)
```
Use `dict[key]` for required config keys.

MartinHjelmare commented on homeassistant/components/sensor/trafikverket_weatherstation.py:
```diff
@@ -4,119 +4,117 @@
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
+async def async_setup_platform(hass, config, async_add_devices,
```
Rename `async_add_devices` to `async_add_entities`.

MartinHjelmare commented on homeassistant/components/sensor/trafikverket_weatherstation.py:
```diff
@@ -4,119 +4,117 @@
+        self._name = SENSOR_TYPES[sensor_type][0]
         self._type = sensor_type
-        self._state = None
+        self._state = STATE_UNKNOWN
```
Init state to `None`.

MartinHjelmare commented on homeassistant/components/sensor/trafikverket_weatherstation.py:
```diff
@@ -4,119 +4,117 @@
-        # air_vs_road contains "Air" or "Road" depending on user input.
-        self._state = final[air_vs_road]["Temp"]
+            self._weather = await\
+                self._weather_api.async_get_weather(self._station)
```
Break the line after the parenthesis instead.

MartinHjelmare commented on homeassistant/components/sensor/trafikverket_weatherstation.py:
```diff
@@ -4,119 +4,117 @@
-        self._state = None
+        self._state = STATE_UNKNOWN
+        self._unit = SENSOR_TYPES[sensor_type][1]
+        self._code = None
```
Not used.

endor-force committed:
Cleaned up based on Martins input

Appreciate the feedback

endor-force commented:
Updated and smoke tested with some scenarios for errors.

[ ] Documentation needs to be updated since it is a change in how to configure
I will fix that within short

MartinHjelmare commented:
Please also add a paragraph in the PR description about the breaking change for the release notes.

MartinHjelmare reviewed (approved):
Ping me when description is updated and docs done.

endor-force commented:
@MartinHjelmare 
I have updated existing PR for the doc and updated the description to accommodate for the changes.

MartinHjelmare commented:
Please write what has changed regarding configuration.

tubalainen commented:
This is sooo great! Thanks guys! Cannot wait for the release! <3
Tack @MartinHjelmare !
